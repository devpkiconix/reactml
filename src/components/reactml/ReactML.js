import React from 'react';
import {
    __, defaultTo, path as pathGet,
    curry,
} from 'ramda';
import mapValues from 'lodash/mapValues';
import dependencies from './dependencies';
import renderLib from './render';
import defaultTagFactory from './materialUiTagFactory';

const renderer = curry(renderLib.render)(dependencies);

const { connect, withStyles } = dependencies;

const defaultToEmpty = defaultTo({})

const state2PropsMaker = (compName, props) => {
    const state2propsPath = ['spec', 'components', compName, 'state-to-props'];
    const stateNodeName = props.spec.state.stateNodeName;

    const oState2pathSpec = defaultToEmpty(
        pathGet(state2propsPath, props)
    );
    return (state) => {
        const imPathGet =
            (dotted) => state.reactml.getIn([stateNodeName].concat(dotted.split('.')));
        const mapped = mapValues(oState2pathSpec, imPathGet);
        // console.log(state.reactml.toJS(), "mapped", mapped);
        return mapped;
    };
};

export const ReactML = (props) => {
    const stateNodeName = props.stateNodeName;
    const compName = props.component;
    const stylesPath = ['spec', 'components', compName, 'styles'];
    const viewNodePath = ['spec', 'components', compName, 'view']
    const mapStateToProps = state2PropsMaker(compName, props);
    const actionExtractor = () => props.actionLib;
    const compProps = {
        tagFactory: props.tagFactory || defaultTagFactory,
        root: pathGet(viewNodePath, props),
        // ...mapStateToProps,
        stateNodeName,
    };
    return React.createElement(
        connect(mapStateToProps, actionExtractor())(
            withStyles(
                defaultToEmpty(
                    pathGet(stylesPath, props)
                ))(
                    renderer
                )
        ),
        compProps);
};

