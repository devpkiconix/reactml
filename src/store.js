import { createStore, applyMiddleware, compose } from 'redux'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import thunk from 'redux-thunk'
import logger from 'redux-logger'
import createHistory from 'history/createBrowserHistory'
import rootReducer from './modules'


export const history = createHistory()

const initialState = {}
const enhancers = []
const middleware = [thunk, routerMiddleware(history)]

if (process.env.NODE_ENV === 'development') {
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__

  middleware.push(logger)
  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension())
  }
}

const composedEnhancers = compose(
  applyMiddleware(...middleware),
  ...enhancers
)


function makeStore() {
  const store = createStore(
    connectRouter(history)(rootReducer),
    initialState,
    composedEnhancers
  );
  return store;
}

export default makeStore;