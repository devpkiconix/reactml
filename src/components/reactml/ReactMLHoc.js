import React from 'react';
import { connect } from 'react-redux';
import { ReactML } from '../../components/reactml/ReactML';

const stateInitHoc = (spec, lib, tagFactory, stateNodeName, view) => {
    class View extends React.Component {
        componentDidMount() {
            if (spec.state.initial) {
                this.props.dispatch({
                    type: 'REACTML_INIT', initial: {
                        ...spec.state.initial,
                    },
                    stateNodeName,
                });
            }
        }

        render() {
            debugger

            return (<ReactML
                tagFactory={tagFactory}
                stateNodeName={stateNodeName}
                spec={spec} component={view}
                actionLib={lib}
            />);
        }
    }

    const mapStateToProps = (state) => { return {} };
    return connect(mapStateToProps)(View);
}


export default stateInitHoc;