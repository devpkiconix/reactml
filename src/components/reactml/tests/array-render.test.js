import React from 'react';
import { Provider } from 'react-redux'
import { ReactML } from '../ReactML';
import ReactMLArrayMapper from '../ArrayMapper';
import ReactMLHoc from '../ReactMLHoc';
import renderer from 'react-test-renderer';
import store from './store';

const tagFactory = {};
const lib = {};

const initial = {
    todoList: [
        { label: "get milk", done: true },
        { label: "buy beans", done: true },
        { label: "grind beans", done: true },
        { label: "brew coffee", done: false },
        { label: "boil milk", done: false },
        { label: "mix'em up", done: false },
        { label: "enjoy", done: false },
    ]
};

describe("Basic reactml rendering", () => {
    beforeEach(() => {
        return store.dispatch({
            type: 'REACTML_INIT',
            initial,
            stateNodeName: 'todoApp'
        });
    });

    it("array rendering", () => {
        const spec = {
            state: {
                stateNodeName: 'todoApp',
                // initial,
            },
            components: {
                todoListView: {
                    'state-to-props': {
                        todoList: 'todoList'
                    },
                    view: {
                        tag: 'div',
                        children: [
                            {
                                tag: 'h4', content: 'Todo List',
                            },
                            {
                                tag: 'ReactMLArrayMapper',
                                props: {
                                    over: '.todoList',
                                    as: 'todo',
                                    component: '...Todo'
                                }
                            }
                        ]
                    }
                }

            }
        };
        const Todo = ({ todo }) =>
            <div className="checkbox">
                <label>
                    <input type="checkbox" value={todo.get('label')}
                        checked={todo.get('done')} />
                    {todo.get('label')}
                </label>
            </div>;

        const actionLib = {};
        const tagFactory = { ReactMLArrayMapper, Todo, };
        const compName = 'todoListView';
        const stateNodeName = 'todoApp';

        const TodoList = ReactMLHoc(spec,
            actionLib, tagFactory,
            stateNodeName, compName);

        const elem = renderer
            .create(
                <Provider store={store}>
                    <TodoList />
                </Provider>);

        const tree = elem.toJSON();
        expect(tree).toMatchSnapshot();
    });
})
