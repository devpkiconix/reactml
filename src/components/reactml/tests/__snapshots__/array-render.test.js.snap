// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic reactml rendering array rendering 1`] = `
<div>
  <h4>
    Todo List
  </h4>
  <div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={true}
          type="checkbox"
          value="get milk"
        />
        get milk
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={true}
          type="checkbox"
          value="buy beans"
        />
        buy beans
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={true}
          type="checkbox"
          value="grind beans"
        />
        grind beans
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={false}
          type="checkbox"
          value="brew coffee"
        />
        brew coffee
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={false}
          type="checkbox"
          value="boil milk"
        />
        boil milk
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={false}
          type="checkbox"
          value="mix'em up"
        />
        mix'em up
      </label>
    </div>
    <div
      className="checkbox"
    >
      <label>
        <input
          checked={false}
          type="checkbox"
          value="enjoy"
        />
        enjoy
      </label>
    </div>
  </div>
</div>
`;
